### PR TITLE
Galois C parser enhancements 

### PR DESCRIPTION
**Pull request by simonjwinwood**
General performance enhancements for the datatype theorems proven by the C parser.

### CODE REVIEW

File: tools/c-parser/PackedTypes.thy [hunk 1]
```diff
@@ -960,4 +960,69 @@ proof (simp add: packed_type_access_ti, rule ext)
   finally show "?LHS = ?RHS" .
 qed
 
+subsection {* Proof automation for packed types *}
+
+definition td_packed :: "'a field_desc typ_desc \<Rightarrow> nat \<Rightarrow> nat \<Rightarrow> bool"
+  where "td_packed t sz al \<longleftrightarrow>
+    td_fafu_idem t \<and> td_fa_hi t \<and> aggregate t \<and> size_td t = sz \<and> align_td t = al"
```
Comment (simonjwinwood) [inline tools/c-parser/PackedTypes.thy]: It is possible that the size_td and align_td calculated here could be used instead of the explicit proofs.

**Comment by lsf37** [comment]
Hey @simonjwinwood and @brianhuffman, sorry it took a while. I have almost everything passing and ready to merge apart from one example C file `tools/autocorres/tests/proof-tests/nested_struct.c` ([here](https://github.com/seL4/l4v/blob/master/tools/autocorres/tests/proof-tests/nested_struct.c)).

The failure is not related to Autocorres (it's before autocorres runs). It's the C parser proof that `s2`, `s3` and `s4` are packed structs that fails. Do you have an idea for a fix? 

The speedup is quite nice! I'm getting 27min wall time for CKernel instead of 39min before.

**Comment by lsf37** [comment]
I spoke too soon: SimplExportAndRefine also fails (times out after 9h, usually take 4h), and AsmRefine fails as well. Both look related to the packed struct issue, though, so I'm hopeful that this is the only underlying cause.

**Comment by simonjwinwood** [comment]
@lsf37 I will take a look, although at this point your guess might be as good as mine :)

**Comment by lsf37** [comment]
Thanks!

**Comment by simonjwinwood** [comment]
Hmm, it works for me?  
```
1587190415: Proving UMM inversion for type s2_C... 
 
1587190415: s2_C: About to size/align...
 
1587190415: s2_C: About to size/align 1...
 
1587190415: s2_C: About to type typ_name ...
 
1587190415: s2_C: About to type/field fl...
 
1587190415: s2_C: About to fg...
 
1587190415: Proving UMM packed type for type s2_C... 
 
1587190415: s2_C: About to type/field fl_Some ...
 
1587190415: s2_C: About to type/field fl_ti ...
 
1587190415: s2_C: About to td_names ...
 
1587190415: s2_C: done
```
This is using 
```
theory nested_struct imports "CParser.CTranslation" begin

external_file "nested_struct.c"
install_C_file "nested_struct.c"
```
Can you try with just my branch?

**Comment by lsf37** [comment]
Hm, I thought I did use your branch. Let me check again.

**Comment by lsf37** [comment]
Strange. I'm still getting the following:
```
1587193928: Proving UMM inversion for type s2_C... 
 
1587193928: s2_C: About to size/align...
 
1587193928: s2_C: About to size/align 1...
 
1587193928: s2_C: About to type typ_name ...
 
1587193928: s2_C: About to type/field fl...
 
1587193928: s2_C: About to fg...
 
1587193928: Proving UMM packed type for type s2_C... 
 
1587193928: Failed to prove UMM packed type for type s2_C
 
1587193928: s2_C: About to type/field fl_Some ...
 
1587193928: s2_C: About to type/field fl_ti ...
 
1587193928: s2_C: About to td_names ...
 
1587193928: s2_C: done
```

This is with
```
theory nested_struct imports "CParser.CTranslation" begin

external_file "nested_struct.c"
install_C_file "nested_struct.c"
```
on revision `c60fdfe15d`.

For `s1` it seems to work:
```
1587193928: Proving UMM inversion for type s1_C... 
 
1587193928: s1_C: About to size/align...
 
1587193928: s1_C: About to size/align 1...
 
1587193928: s1_C: About to type typ_name ...
 
1587193928: s1_C: About to type/field fl...
 
1587193928: s1_C: About to fg...
 
1587193928: Proving UMM packed type for type s1_C... 
 
1587193928: s1_C: About to type/field fl_Some ...
 
1587193928: s1_C: About to type/field fl_ti ...
 
1587193928: s1_C: About to td_names ...
 
1587193928: s1_C: done
```

**Comment by lsf37** [comment]
Could there be any influence from env variables being set/not set? I usually have SORRY_MODIFIES_PROOFS on for my laptop. Maybe that produces a side effect somewhere.

**Comment by lsf37** [comment]
I will see if I can run it on a different machine, but that will take a bit to set up.

**Comment by lsf37** [comment]
Getting the same failure on the server. Any tips on debugging it, any output enrichment I could provide?

**Comment by simonjwinwood** [comment]
Wrong branch :/  Looking into it

**Comment by simonjwinwood** [comment]
OK, fixed that issue, we were missing array lemmas

**Comment by lsf37** [comment]
Cool, I'm pulling and testing. You can ignore the DCO on the PR, I can sign off the commits for you since you have a CLA in place.

**Comment by lsf37** [comment]
Cool, I can confirm that this one now works, including the full AutoCorresTest session. Thanks for the fix!

I'm doing a full run of everything which will take 6-ish hours probably. If that passes I'll rebase and merge tomorrow.

**Comment by simonjwinwood** [comment]
OK, let me know and I will rebase the heap tactic patch and submit another PR

**Comment by lsf37** [comment]
Sadly it only fixed the first session failure, not the other two. The session `AsmRefine` is still relatively easy to build (directly on `CParser`), the session `SimplExportAndRefine` needs `CSpec`. I'll paste the details for `AsmRefine` here and dig a bit more into `SimplExportAndRefine` in the meantime.

For `AsmRefine` the failure is in `tools/asmrefine/testfiles/global_array_swap_gref.thy`, and it looks like the the tactic there is expecting the following term to simplify:
```isabelle
 of_nat
   ((align_of TYPE(struct_blob_C[192]) - 44 mod align_of TYPE(struct_blob_C[192])) mod align_of TYPE(struct_blob_C[192]))
```

The tactic `ptr_simp` that seems to be the one that is ultimately failing  is in `tools/asmrefine/SimplExport`:
```sml
 fun ptr_simp ctxt = ctxt addsimps @{thms CTypesDefs.ptr_add_def size_of_def size_td_array
        field_lvalue_offset_eq align_td_array' word_of_int scast_def[symmetric]
        ucast_def[symmetric]
        sint_sbintrunc' word_smod_numerals word_sdiv_numerals sdiv_int_def smod_int_def}
  |> Simplifier.rewrite
```

Overall the tactic is translating all global variables to be part of the heap instead of separate, as a first step in the binary verification toolchains (basically doing a few first compilation steps inside Isabelle).

The full message is:
```isabelle
exception TERM raised⌂:
  emit_func_body: add_a_thing: convert_upd_ph3: ptr_simp_term: of_nat
  TRY
    IF ´data_num <s 0xC0 THEN
      Guard ArrayBounds ⦃´data_num <s 0xC0⦄
       (Guard ArrayBounds ⦃0 <=s ´data_num⦄
         (Guard C_Guard ⦃c_guard (things_C ´toplevel +⇩p sint ´thing_num)⦄
           (´globals :==
              t_hrs_'_update
               (hrs_mem_update
                 (heap_update
                   (PTR(32 word)
                     &(PTR_COERCE(struct_blob_C[192] → struct_blob_C)
                        (PTR(struct_blob_C[192]) &(things_C ´toplevel +⇩p sint ´thing_num→[''list_of_data_C''])) +⇩p
                       sint ´data_num→[''index_C'']))
                   (SCAST(32 signed → 32) 1))))))
    ELSE
      Guard C_Guard ⦃c_guard (things_C ´toplevel +⇩p sint ´thing_num)⦄
       (Guard SignedArithmetic ⦃- 2147483648 ≤ - sint 1 ∧ - sint 1 ≤ 2147483647⦄
         (´globals :==
            t_hrs_'_update
             (hrs_mem_update
               (heap_update
                 (PTR(32 word)
                   &(PTR(range_blob_C) &(things_C ´toplevel +⇩p sint ´thing_num→[''range_in_stuff_C''])→[''end_C'']))
                 (SCAST(32 signed → 32) (- 1))))))
    FI
  CATCH SKIP
  END
  True
  heap_update
   (PTR(32 word)
     &(PTR_COERCE(struct_blob_C[192] → struct_blob_C)
        (PTR(struct_blob_C[192])
          &(h_val (pseudo_acc Mem) (PTR(AnonStruct1' ptr) (symbol_table ''toplevel'' + 0x10)) +⇩p
            sint (pseudo_acc thing_num___int#v)→[''list_of_data_C''])) +⇩p
       sint (pseudo_acc data_num___int#v)→[''index_C'']))
   (SCAST(32 signed → 32) 1) (pseudo_acc Mem)
  of_nat
   ((align_of TYPE(struct_blob_C[192]) - 44 mod align_of TYPE(struct_blob_C[192])) mod align_of TYPE(struct_blob_C[192]))
```

**Comment by lsf37** [comment]
At least some good(ish :-)) news: the failure in `SimplExportAndRefine` is of the same form. It wants to simplify:
```isabelle
 of_nat ((align_of TYPE(mdb_node_C) - 8 mod align_of TYPE(mdb_node_C)) mod align_of TYPE(mdb_node_C))
```

Full message is:
```isabelle
ptr_val (pseudo_acc p___ptr_to_unsigned_char#v) - ptr_val (pseudo_acc q___ptr_to_unsigned_char#v) ≡
ptr_val (pseudo_acc p___ptr_to_unsigned_char#v) - ptr_val (pseudo_acc q___ptr_to_unsigned_char#v) 
ptr_val (pseudo_acc recvBuf___ptr_to_unsigned_long#v +⇩p uint (pseudo_acc i___unsigned_long#v + SCAST(32 signed → 32) 1)) -
ptr_val (pseudo_acc sendBuf___ptr_to_unsigned_long#v +⇩p uint (pseudo_acc i___unsigned_long#v + SCAST(32 signed → 32) 1)) ≡
ptr_val (pseudo_acc recvBuf___ptr_to_unsigned_long#v) - ptr_val (pseudo_acc sendBuf___ptr_to_unsigned_long#v) 
val it = (): unit
val CFunDump_filename = "ARM/CFunDump.txt": string 
exception TERM raised (line 1112 of "~/verification/l4v/tools/asmrefine/SimplExport.thy"):
  emit_func_body: cteMove: emit_body call: convert_upd_ph3: ptr_simp_term: of_nat
  TRY
    lvar_nondet_init mdb_' mdb_'_update;;
    lvar_nondet_init prev_ptr_' prev_ptr_'_update;;
    lvar_nondet_init next_ptr_' next_ptr_'_update;;
    SKIP;;
    SKIP;;
    Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t ´srcSlot⦄
     (´mdb :== h_val (hrs_mem ´t_hrs) (PTR(mdb_node_C) &(´srcSlot→[''cteMDBNode_C''])));;
    Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t ´destSlot⦄
     (´globals :== t_hrs_'_update (hrs_mem_update (heap_update (PTR(cap_C) &(´destSlot→[''cap_C''])) ´newCap)));;
    ´ret__struct_cap_C :== CALL cap_null_cap_new_'proc();;
    Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t ´srcSlot⦄
     (´globals :== t_hrs_'_update (hrs_mem_update (heap_update (PTR(cap_C) &(´srcSlot→[''cap_C''])) ´ret__struct_cap_C)));;
    Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t ´destSlot⦄
     (´globals :== t_hrs_'_update (hrs_mem_update (heap_update (PTR(mdb_node_C) &(´destSlot→[''cteMDBNode_C''])) ´mdb)));;
    ´ret__struct_mdb_node_C :== CALL mdb_node_new_'proc(SCAST(32 signed → 32) 0,SCAST(32 signed → 32) false,
    SCAST(32 signed → 32) false,SCAST(32 signed → 32) 0);;
    Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t ´srcSlot⦄
     (´globals :==
        t_hrs_'_update
         (hrs_mem_update (heap_update (PTR(mdb_node_C) &(´srcSlot→[''cteMDBNode_C''])) ´ret__struct_mdb_node_C)));;
    ´ret__unsigned :== CALL mdb_node_get_mdbPrev_'proc(´mdb);;
    ´prev_ptr :== UCAST(32 → 32) ´ret__unsigned;;
    IF ´prev_ptr ≠ 0 THEN
      Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t PTR(cte_C) ´prev_ptr⦄
       (CALL mdb_node_ptr_set_mdbNext_'proc(PTR(mdb_node_C) &(PTR(cte_C) ´prev_ptr→[''cteMDBNode_C'']),
        UCAST(32 → 32) (ptr_val ´destSlot)))
    FI;;
    ´ret__unsigned :== CALL mdb_node_get_mdbNext_'proc(´mdb);;
    ´next_ptr :== UCAST(32 → 32) ´ret__unsigned;;
    IF ´next_ptr ≠ 0 THEN
      Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t PTR(cte_C) ´next_ptr⦄
       (CALL mdb_node_ptr_set_mdbPrev_'proc(PTR(mdb_node_C) &(PTR(cte_C) ´next_ptr→[''cteMDBNode_C'']),
        UCAST(32 → 32) (ptr_val ´destSlot)))
    FI
  CATCH SKIP
  END
  True
  IF ´next_ptr ≠ 0 THEN
    Guard C_Guard ⦃hrs_htd ´t_hrs ⊨⇩t PTR(cte_C) ´next_ptr⦄
     (CALL mdb_node_ptr_set_mdbPrev_'proc(PTR(mdb_node_C) &(PTR(cte_C) ´next_ptr→[''cteMDBNode_C'']),
      UCAST(32 → 32) (ptr_val ´destSlot)))
  FI
  λs. v32_'_update (λ_. UCAST(32 → 32) (ptr_val (destSlot_' s)))
       (mdb_node_ptr_'_update (λ_. PTR(mdb_node_C) &(PTR(cte_C) (next_ptr_' s)→[''cteMDBNode_C''])) s)
  PTR(mdb_node_C) &(PTR(cte_C) (pseudo_acc next_ptr___unsigned_long#v)→[''cteMDBNode_C''])
  of_nat ((align_of TYPE(mdb_node_C) - 8 mod align_of TYPE(mdb_node_C)) mod align_of TYPE(mdb_node_C))
```

**Comment by simonjwinwood** [comment]
Hmm, I was hoping it was just a missing array thing, but the second rules that out.  It might be that the align_of lemmas are no longer being put into the global simpset?  I will take a look.

**Comment by simonjwinwood** [comment]
It looks like the first is fixed if align_of_def is in added to the simpset, I wonder if we forgot some lemmas?

**Comment by lsf37** [comment]
If I try to simplify `align_of TYPE(mdb_node_C)` in isolation before your patch, I get no progress with just simp, `simp add: align_of_def` works, though, as you say.

Maybe the tactic is expecting a particular name or form for `align_of` theorems that is now different?

**Comment by lsf37** [comment]
Conversely, it's of course possible that it now does more than before and therefore find itself in an unexpected state

**Comment by simonjwinwood** [comment]
It actually has to do with an earlier step (same ptr_simp_term) when simplifying
```
 &((p :: AnonStruct1' ptr)→[''list_of_data_C''])"
```

**Comment by simonjwinwood** [comment]
I tracked it down to the new _fl rules, in particular
```
AnonStruct1'_list_of_data_C_fl
```
which differs.  This patch changes how these lemmas are stated (basically removed unneeded generality) which makes them simpler, but also changes the defn.

Previously
```
field_lookup (typ_info_t TYPE(AnonStruct1')) (''list_of_data_C'' # ?fs) ?m ≡
field_lookup (adjust_ti (typ_info_t TYPE(struct_blob_C[192])) list_of_data_C (list_of_data_C_update ∘ (λx _. x))) ?fs
 (44 + ?m)
```
and now
```
field_lookup (typ_info_t TYPE(AnonStruct1')) [''list_of_data_C''] 0 =
Some
 (adjust_ti (typ_info_t TYPE(struct_blob_C[192])) list_of_data_C (list_of_data_C_update ∘ (λx _. x)),
  (align_of TYPE(struct_blob_C[192]) - 44 mod align_of TYPE(struct_blob_C[192])) mod
  align_of TYPE(struct_blob_C[192]) +
  44)
```
I think this is a net improvement, but it may break some scripts.  There is a bug where it isn't simplifying enough, which I will fix.

**Comment by lsf37** [comment]
Ok, got it. Specifically it introduces the `align_of` terms that the script didn't expect. Nice work tracking that one down. I agree that it's an improvement, and if we can get `SimplExport` going again I'm happy with it. It's possible that the tactic just needs an `align_of_def` in its simpset to solve these.

**Comment by simonjwinwood** [comment]
I fixed the _fl generation, it was missing some simp rules.

**Comment by simonjwinwood** [comment]
Let me know if this is actually faster :)

**Comment by lsf37** [comment]
Thanks, will pull in and do another run!

**Comment by lsf37** [comment]
So, `AsmRefine` passes, but `SimplExportAndRefine` (which basically just runs `AsmRefine` on seL4) still has a problem. It passes the previous issue, but ends with
```
success on Kernel_C.thread_state_ptr_get_blockingIPCCanGrantReply  [6.788s elapsed time, 22.201s cpu time, 0.929s GC time]
success on Kernel_C.mdb_node_ptr_mset_mdbNext_mdbRevocable_mdbFirstBadged  [5.405s elapsed time, 16.669s cpu time, 0.693s GC time]
success on Kernel_C.thread_state_ptr_set_blockingIPCCanGrantReply  [10.155s elapsed time, 22.639s cpu time, 0.912s GC time]
### theory "SimplExportAndRefine.SEL4GraphRefine"
### 4233.918s elapsed time, 15267.279s cpu time, 985.842s GC time
### No traces for session: SimplExportAndRefine
*** Failures! Check the `#failures` field of the debug parameter. [4225.072s elapsed time, 15256.839s cpu time, 983.680s GC time]
*** At command "ML" (line 73 of "/home/bamboo/bamboo-agent-home/xml-data/build-dir/L4V-TESTBOARD-RISCV64VERIFY/l4v/proof/asmrefine/SEL4GraphRefine.thy")
Unfinished session(s): SimplExportAndRefine
```
That's obviously not very helpful, I'll have to go in and see if I can find out more interactively and/or ask some of our binary verification experts to join in. Will let you know!

**Comment by simonjwinwood** [comment]
Argh, it might be less work to split the patch into the various speedups and see what is breaking.  Is it any faster with the other sessions (or haven't you gotten that far yet?)

**Comment by lsf37** [comment]
Yes, speedup is there, almost as good as before. I'm still keen :-)

``
 Finished CKernel                   passed      ( 0:28:19 real,  1:24:04 cpu, 17.55GB)
``
compared to
```
Finished CKernel                   passed      ( 0:37:56 real,  1:33:06 cpu, 17.50GB)
```
on master. Interesting that it claims to use a tiny bit more memory, but the memory numbers are not that reliable in any case.

Bisecting per patch is a good idea, I'll give that go!

**Comment by simonjwinwood** [comment]
Cool, that is a reasonable speedup.  I think if you used a sum type instead of strings (would need a bit of reengineering though) things could be faster.

**Comment by lsf37** [comment]
> I think if you used a sum type instead of strings (would need a bit of reengineering though) things could be faster.

That topic has come up a few times, and I think you're right.

One of the nice things of Strings is that you can write rules about names before the type for names exist (basically, it's the universe type for names). You could make that a type parameter, though, maybe in a class that gives you a bunch of operations on them. Would indeed take a bit of reengineering, and probably updates in the proofs themselves, but if you're interested, we could have a look at that together over time.

**Comment by lsf37** [comment]
(Bisect is running, let's see what we get)

**Comment by simonjwinwood** [comment]
One thing is that the last few patches were to fix bugs in both the packed and _fl thms

**Comment by lsf37** [comment]
Should we sqash/rebase first?

**Comment by simonjwinwood** [comment]
> if you're interested, we could have a look at that together over time.

I don't have umpteemth million lines of proof though :)  You could probably cherry pick the relevant commits, but if you could get something more informative for the breakage it might also help.

**Comment by lsf37** [comment]
Ok, results from bisect: 177667e11320dca3467844d2ed0c4e2830864d13 works, and 439a1691bfa72da990271db2266e433c209932e9 is the first that breaks. It does break on the known array issue, though. So it would make sense to try and split+merge the fixes into the patches they address, then we can possibly pin it down a bit more.

**Comment by lsf37** [comment]
I can attempt that if you want, but it's probably better if you do it yourself because you've written the fixes ;-)

**Comment by lsf37** [comment]
I've combined 4a69ae233d0fd983b1e87b2cf7977d91a8c7fe92 into 439a1691bfa72da990271db2266e433c209932e9 and 5c69fcfbda2ca9947eac1ab407fc554685224966 into e46f626d1800bd9f7313e2375bb38f5ef38dea86 and rebased that.

The result is in https://github.com/seL4/l4v/commits/galois-parser of which f540615bf095194fa3ef67e2765eddb11b3d9392 is confirmed working and 668c4cbcb747a8e339066215207ef2a56c8d293a is currently running. Will report back on that.

**Comment by simonjwinwood** [comment]
OK, not that surprising that the more radical change broke things.  I would be surprised if the extra generality is used, but it might be the problem?

**Comment by lsf37** [comment]
668c4cb also works! So 6d2ff215ffc371fd451d2cf0971d0feaf91fdce9 is the one that fails.

**Comment by lsf37** [comment]
Not sure if that is enough information to do anything, but one thing we could do is merge those two first and look at the last one in more detail in a separate PR.

**Comment by lsf37** [comment]
The two working patches already give a reasonable speedup (~30min instead of the 28min that was the best, both down from 39min for CKernel).

**Comment by simonjwinwood** [comment]
Hmm, OK, it would be interesting to see what broke, but might not be worth the extra debugging time to include the other things.  I wonder if the new _fl proofs give better proofs in use?

**Comment by lsf37** [comment]
They might, and I'm still keen to look into it. It'd just be nice to already have the ones that work merged as solid progress, and also I'll have to wait less long for CKernel to build  :-). We're currently working hard on finishing CRefine for RISC-V, so any reduction helps. 

Do you want to rebase yourself or are you happy for me to do that? If you want to do it yourself, we should wait one more cycle, because we managed to break master for a bit, which stops pushing `l4v` out to github. Should be fixed again already, but will take a few hours for the test to complete and the deployment to run.

**Comment by simonjwinwood** [comment]
Your merge it, it is your codebase :)

**Comment by lsf37** [comment]
Happy to :-)

**Comment by lsf37** [comment]
Rebased and merged at 3c3514fe99082f7b6a6fb8445b8dfc592ff7f02b! There'll be conflicts on this branch now, because I updated some minor whitespace and style things. Might see if I can fix those on github directly.

**Comment by lsf37** [comment]
That seems to have worked. The diff shows only the `field_lookup` commit and its fixup. Will have a closer look at this last one next week, potentially the week after (big deadline on Fri).

**Comment by simonjwinwood** [comment]
I think the heap tactic tool relies on that form of lemma, but it should be reasonably easy to change if we don't include it.

**Comment by lsf37** [comment]
Given that we haven't made any progress on this in the past few years and we did get the main speedup going from the parts that are merged, I'm closing this PR.

If anyone wants to take a stab at reviving it and updating the failing tactics, do feel free to re-open!

**Pull request closed without merging**